sm/plugins/tedge_apt_plugin/src/main.rs

```diff
@@ -0,0 +1,138 @@
+    },
+
+    /// Prepare a sequences of install/remove commands
+    Prepare,
+
+    /// Finalize a sequences of install/remove commands
+    Finalize,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum InternalError {
+    #[error("Fail to run `{cmd}`: {from}")]
+    ExecError { cmd: String, from: std::io::Error },
+}
+
+impl InternalError {
+    pub fn exec_error(cmd: impl Into<String>, from: std::io::Error) -> InternalError {
+        InternalError::ExecError {
+            cmd: cmd.into(),
+            from,
+        }
+    }
+}
+
+fn run(operation: PluginOp) -> Result<std::process::ExitStatus, InternalError> {
+    let status = match operation {
+        PluginOp::List {} => {
+            let apt = Command::new("apt")
+                .args(vec!["--installed", "list"])
+                .stdout(Stdio::piped()) // To pipe apt.stdout into grep.stdin
+                .spawn()
+                .map_err(|err| InternalError::exec_error("apt", err))?;
+
+            let grep = Command::new("grep")
+                .args(vec!["-v", "automatic"])
+                .stdin(apt.stdout.unwrap()) // Cannot panics: apt.stdout has been set
+                .stdout(Stdio::piped()) // To pipe grep.stdout into awk.stdin
+                .spawn()
+                .map_err(|err| InternalError::exec_error("grep", err))?;
+
+            // apt output    = openssl/focal-security,now 1.1.1f-1ubuntu2.3 amd64 [installed]
+            // awk -F '[/ ]' =   $1   ^       $2         ^   $3            ^   $4
+            // awk print     =   name ^                  ^   version       ^
+            let status = Command::new("awk")
+                .args(vec![
+                    "-F",
+                    "[/ ]",
+                    r#"{print "{\"name\":\""$1"\",\"version\":\""$3"\"}"}"#,
+                ])
+                .stdin(grep.stdout.unwrap()) // Cannot panics: grep.stdout has been set
+                .status()
+                .map_err(|err| InternalError::exec_error("awk", err))?;
+
+            status
+        }
+
+        PluginOp::Install { module, version } => {
+            if let Some(version) = version {
+                run_cmd(
+                    "apt-get",
+                    &format!("install --quiet --yes {}={}", module, version),
+                )?
+            } else {
+                run_cmd("apt-get", &format!("install --quiet --yes {}", module))?
+            }
+        }
+
+        PluginOp::Remove {
+            module,
+            version: _unused,
+        } => run_cmd("apt-get", &format!("remove --quiet --yes {}", module))?,
+
+        PluginOp::Prepare => run_cmd("apt-get", &format!("update --quiet --yes"))?,
+
+        PluginOp::Finalize => run_cmd("apt-get", &format!("auto-remove --quiet --yes"))?,
+    };
+
+    Ok(status)
+}
+
+fn run_cmd(cmd: &str, args: &str) -> Result<std::process::ExitStatus, InternalError> {
+    let args: Vec<&str> = args.split_whitespace().collect();
+    let status = Command::new(cmd)
```

Comment:
Should there be some sort of timeout to prevent command executions taking too long?

Reply:
I think this should be the responsibility of the `sm-agent`. So this will be done in a single place.